src/Microsoft.VisualStudio.Editors/MyApplication/MyApplicationProjectLifetimeTracker.vb

```diff
@@ -0,0 +1,135 @@
+﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+Imports System.Collections.Immutable
+Imports Microsoft.VisualStudio.Editors.Interop
+Imports Microsoft.VisualStudio.Shell
+Imports Microsoft.VisualStudio.Shell.Interop
+
+Namespace Microsoft.VisualStudio.Editors.MyApplication
+    Friend Class MyApplicationProjectLifetimeTracker
+        Implements IVsRunningDocTableEvents
+        Implements IVsSolutionEvents
+
+        Private Shared ReadOnly s_instance As MyApplicationProjectLifetimeTracker = New MyApplicationProjectLifetimeTracker
+
+        Private _managerInstances As ImmutableDictionary(Of UInteger, MyApplicationProperties) = ImmutableDictionary(Of UInteger, MyApplicationProperties).Empty
+        Private ReadOnly _rdt As RunningDocumentTable
+        Private ReadOnly _solution As IVsSolution
+
+        Private Sub New()
+            _rdt = New RunningDocumentTable
+            _rdt.Advise(Me)
+
+            _solution = DirectCast(ServiceProvider.GlobalProvider.GetService(GetType(IVsSolution)), IVsSolution)
+            Dim solutionCookie As UInteger
+            _solution.AdviseSolutionEvents(Me, solutionCookie)
+        End Sub
+
+        Public Shared Function Track(projectHierarchy As IVsHierarchy) As MyApplicationProperties
+            Return s_instance.TrackInternal(projectHierarchy)
+        End Function
+
+        Private Function TrackInternal(projectHierarchy As IVsHierarchy) As MyApplicationProperties
+            Dim cookie = GetProjectFileCookie(projectHierarchy)
+
+            ' Because GetOrAdd always creates a new value, even if using a factory, we just create an object and only
```

Comment:
I could have sworn I looked at this, and saw that valueFactory was always called, but looking now you're right it isn't called if the key exists. I guess I could change this to use a Func then, but I'd still want to keep the `Init` call out of it just in case, as that does a bunch of interop stuff and calls into the hierarchy that its worth avoiding IMO. I could also change this to use a `Lazy`.. too many options 😛

Reply:
Changed this to a lambda, and corrected the comment.